### ISSUE TITLE
Support opening existing .etl files

### ISSUE THREAD

**Issue by jrmuizel** [n4r1b/ferrisetw#7]
This would be quite useful.

See also: https://github.com/microsoft/krabsetw/issues/160

**Comment by jrmuizel** [comment]
I've put a rough fork that I'm using for this at https://github.com/jrmuizel/etw-profiler/tree/main/etw-reader

**Comment by n4r1b** [comment]
Hey @jrmuizel ! Nice one, That's indeed quite useful!!

I would love if you'd like to adapt your fork to `ferrisetw` and open a PR with the changes, I would gladly check it and merge it. If that's not possible I'll try to do it myself asap 😃

**Comment by jrmuizel** [comment]
I'm not sure I'll have time to submit PRs with my changes, but I'll try to at least open some issues about the things that I changed.

**Comment by daladim** [comment]
I've just implemented support for _dumping_ into ETL files in #91.

I won't be able to implement _reading_ from ETL files for now.
I tried to do so, but that's harder as I planned, because ferrisetw has too much coupling between Providers (that make only sense for real-time traces) to callbacks (that make sense for both real-time and file-based traces).
That's good because this design makes it easy to add multiple callbacks for multiple ETW providers, and call the right callbacks for the right events. But this prevents us from easily call `OpenTrace` without the `StartTrace`/`EnableTraceEx2` (which are only required for real-time sessions).

I'm afraid this would require some refacto. Doing it well would be rather heavy (and I won't have time for it in the short run). Also, this may require to change the exposed, public API and force releasing a new major version.
